GH-3336: Change MongoDb Store sequence to long

artembilan opened the pull request:
Fixes https://github.com/spring-projects/spring-integration/issues/3336

Turns out there are some scenarios where too many messages
are transferred through the message store, so `int` for
sequence is not enough as a type

* Change sequence to `long` to widen a sequence lifespan

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-3336: Change MongoDb Store sequence to long

Fixes https://github.com/spring-projects/spring-integration/issues/3336

Turns out there are some scenarios where too many messages
are transferred through the message store, so `int` for
sequence is not enough as a type

* Change sequence to `long` to widen a sequence lifespan

garyrussell commented:
What's new, migration? Breaking change? (i.e. can documents created with int be read as long?). If not, we probably need to provide a conversion utility.

artembilan commented:
Sure! I'll prepare Migration Guide note.
Probably the best way to go is to remove an existing `messagesSequence` document and let the framework to recreate it with the `long` type already.

I don't think there is the problem with casting existing `int` values for allocated sequences in message documents to `long`...

artembilan committed:
* Change MongoDb store to deal with `Number.longValue()`
instead of casting which doesn't work from `Integer` to `Long`.
This way we can keep an old sequence document with an `int`
type for value
* Documents with new `long` type for their sequence field are OK.
The `NumberToNumberConverter` has an effect converting `int` to `long`
properly.

artembilan commented:
Pushed some fix to keep compatibility with the previous MongoDb collection state.

The Migration Guide note still has to be done.
